feat: Checkpoint Acquisition Layer — Local, HTTP, S3 sources (#458)

JesperDramsch opened the pull request:
**Part of the checkpoint system refactor (#248).** Implements the Checkpoint Acquisition Layer as native `PipelineStage` classes built on the Phase 1 infrastructure (PR #501).

> ⚠️ **Complete reimplementation.** The original PR #464 code (sync `CheckpointSource.fetch()` interface) has been replaced with a fresh implementation using Phase 1's async `PipelineStage.process(context)` pattern. The old review findings are tracked

### What this PR does

Adds three checkpoint source implementations that fetch checkpoints from different backends and populate `CheckpointContext.checkpoint_data` for downstream loading strategies:

| Source | What it does | Key dependency |
|--------|-------------|----------------|
| `LocalSource` | Loads from local filesystem | `torch.load()` + Phase 1 `detect_checkpoint_format()` |
| `HTTPSource` | Async download with retry | Phase 1 `download_with_retry()` (aiohttp, exponential backoff) |
| `S3Source` | Download from S3-compatible storage | `boto3` (lazy import, credential delegation) |

All three extend `CheckpointSource(PipelineStage)` and implement `async process(context) -> context`.

### Architecture

```
CheckpointContext enters with checkpoint_path or config["url"]
        ↓
  Source stage loads/downloads checkpoint
        ↓
CheckpointContext exits with checkpoint_data populated + checkpoint_format detected
```

Data flows through `CheckpointContext`, not return values. Each source is a standalone pipeline stage — no shared dispatcher, no registry routing.

### Design decisions

- **Reimplementation, not adaptation**: PR #458's download logic is 5–15 lines per source. The registries and class hierarchy from #458 are replaced by Phase 1's `ComponentCatalog` and `CheckpointPipeline`. Wrapping 600-line classes to reuse 15 lines would add needless indirection.
- **Local + HTTP + S3 only**: GCS and Azure deferred per review feedback (Ana: remove cloud provider refs). Trivial to add later — ~10 lines each, same pattern.
- **Phase 1 error types everywhere**: `CheckpointNotFoundError`, `CheckpointSourceError` — never stdlib exceptions.
- **S3 credential delegation**: Entirely boto3's default chain. Works across AWS, EWC, MinIO, Ceph without code changes.
- **S3 error mapping**: Every boto3 error code mapped to actionable messages with Phase 1 exception types.
- **Lazy imports**: `boto3` imported inside method, not at module level. `ImportError` includes pip install command.

### What's NOT in this PR

| Deferred | Why | Tracked in |
|----------|-----|-----------|
| GCS source | Ana review feedback — remove cloud refs | Follow-up PR |
| Azure source | Same | Follow-up PR |
| MLFlow source | Phase 3 backwards compat | #495 |
| Checkpoint caching | Architecture supports it, not needed yet | Follow-up PR |
| Progress reporting | Phase 1 has hooks for it | Follow-up PR |
| safetensors format | Phase 1 enum doesn't include it | Follow-up issue |

### Test coverage

151+ tests passing, 5 skipped (network-dependent, gated by `ANEMOI_TEST_NETWORK`):

```
training/tests/checkpoint/sources/
├── test_base.py      — 3 tests (inheritance, abstract, interface)
├── test_local.py     — ~30 tests (load, format detect, errors, edge cases)
├── test_http.py      — ~30 tests (download_with_retry usage, error wrapping)
└── test_s3.py        — ~30 tests (URL parsing, error mapping, lazy import, cleanup)
```

Test philosophy: real-behavior tests for LocalSource (zero mocks, actual torch.save/load), mock-only for network sources (boto3, aiohttp).

### How to review

```bash
# Run the full source test suite
pytest training/tests/checkpoint/sources/ -v --tb=short

# Check no Phase 1 files were modified
git diff main -- training/src/anemoi/training/checkpoint/base.py \
  training/src/anemoi/training/checkpoint/pipeline.py \
  training/src/anemoi/training/checkpoint/exceptions.py
```

### Try it yourself

```bash
git checkout feature/checkpoint-sources
cd training
pip install -e .

# Load a local checkpoint through the pipeline
python -c "
import asyncio, torch
from pathlib import Path
from anemoi.training.checkpoint.base import CheckpointContext
from anemoi.training.checkpoint.sources.local import LocalSource

# Create a dummy checkpoint
torch.save({'state_dict': {'layer.weight': torch.randn(10, 5)}, 'epoch': 3}, '/tmp/test.ckpt')

# Run it through the source stage
source = LocalSource()
ctx = CheckpointContext(checkpoint_path=Path('/tmp/test.ckpt'))
result = asyncio.run(source.process(ctx))

print(f'Format:  {result.checkpoint_format}')
print(f'Keys:    {list(result.checkpoint_data.keys())}')
print(f'Weights: {result.checkpoint_data[\"state_dict\"][\"layer.weight\"].shape}')
"
```

For S3 (if you have credentials):
```bash
python -c "
import asyncio
from anemoi.training.checkpoint.base import CheckpointContext
from anemoi.training.checkpoint.sources.s3 import S3Source

source = S3Source()
ctx = CheckpointContext(config={'url': 's3://your-bucket/your-checkpoint.ckpt'})
result = asyncio.run(source.process(ctx))
print(f'Loaded {len(result.checkpoint_data[\"state_dict\"])} parameters from S3')
"
```

### Phase 1 note

6 Phase 1 files have bugfixes on this branch (283 insertions, 137 deletions). Key dependency: `formats.py:detect_format_from_data()` is called by `sources/base.py`. These fixes are also cherry-picked onto the loading-orchestration branch. Pragmatic approach: include fixes in whichever PR merges first, the other gets a clean rebase.

**Closes #458**
**Part of #248**

---

<details>
<summary>Original PR description (archived 2026-03-20)</summary>

## Description

This PR introduces a comprehensive checkpoint loading system that supports loading model checkpoints from various sources including local files, S3, HTTP/HTTPS, Google Cloud Storage, and Azure Blob Storage. The implementation provides a modular, extensible architecture that separates checkpoint retrieval from model loading strategies and includes robust error handling and validation.

## What problem does this change solve?

This change addresses several key problems in the current checkpoint loading workflow:

**New Feature**: Adds multi-source checkpoint loading capabilities with support for:
- **Local filesystem**: Standard file path loading
- **HTTP/HTTPS**: Download from web URLs  
- **Amazon S3**: S3 bucket support with boto3
- **Google Cloud Storage**: GCS support with google-cloud-storage
- **Azure Blob Storage**: Azure support with azure-storage-blob

**Enhanced Model Loading Strategies**:
- **StandardModelLoader**: Standard PyTorch Lightning checkpoint loading
- **TransferLearningModelLoader**: Handle size mismatches and partial weight loading
- **WeightsOnlyModelLoader**: Load only model weights, skip optimiser states

**Architecture Improvements**:
- **Separation of concerns**: Checkpoint retrieval vs model loading logic
- **Extensibility**: Easy to add new sources or loading strategies via registry pattern
- **Testability**: Mock-friendly design with clear interfaces
- **Error handling**: Informative error messages and graceful failures
- **Performance**: Efficient temporary file handling and cleanup

## What issue or task does this change relate to?

**Closes #458**

Should be updated and merged after #410 / #422 

## Additional notes

### Implementation Details

The PR introduces two main components:

**CheckpointLoaders (`checkpoint_loaders.py`)**:
- Abstract base class `CheckpointLoader` with pluggable implementations
- `LocalCheckpointLoader` for filesystem access
- `RemoteCheckpointLoader` with multi-cloud support
- Automatic source detection and loader selection
- Comprehensive error handling with clear messages

**ModelLoading (`model_loading.py`)**:
- Abstract base class `ModelLoader` for different loading strategies
- Standard Lightning checkpoint compatibility
- Transfer learning with size mismatch handling
- Metadata preservation and data_indices validation
- Registry pattern for extensibility

### Usage Examples

```python
# Load from any supported source
checkpoint = load_checkpoint_from_source("s3://bucket/model.ckpt")
checkpoint = load_checkpoint_from_source("https://example.com/model.ckpt")
checkpoint = load_checkpoint_from_source("/local/path/model.ckpt")

# Load model with transfer learning
model = load_model_from_checkpoint(
    model=my_model,
    checkpoint_source="s3://bucket/pretrained.ckpt",
    loader_type="transfer_learning",
    skip_mismatched=True
)

# Register custom loaders
registry.register("custom_source", CustomCheckpointLoader())
registry.register("custom_strategy", CustomModelLoader())
```

### Testing and Quality Assurance

- **100% line coverage** across all loaders and strategies
- **Integration tests** for multi-source scenarios
- **Error handling tests** for edge cases and failures
- **Mock-based testing** to avoid external dependencies

```bash
# Run all checkpoint loading tests
pytest training/tests/utils/test_checkpoint_loaders.py -v
pytest training/tests/utils/test_model_loading.py -v

# Test with different sources (requires cloud credentials)
pytest training/tests/utils/ -k "remote" --cloud-tests
```

### Compatibility and Dependencies

- **Dependencies**: Optional cloud dependencies (boto3, google-cloud-storage, azure-storage-blob)
- **Type safety**: Full type annotations and strict typing
- **Code quality**: Passes all linting rules (ruff, black, isort)

### Future Extensions

This architecture makes it easy to add:
- Additional cloud providers (OCI, DigitalOcean Spaces, etc.)
- Custom authentication mechanisms
- Caching and optimisation strategies
- Checkpoint validation and integrity checks
- Progress tracking for large downloads

</details>

<!-- readthedocs-preview anemoi-training start -->
----
📚 Documentation preview 📚: https://anemoi-training--464.org.readthedocs.build/en/464/

<!-- readthedocs-preview anemoi-training end -->

<!-- readthedocs-preview anemoi-graphs start -->
----
📚 Documentation preview 📚: https://anemoi-graphs--464.org.readthedocs.build/en/464/

<!-- readthedocs-preview anemoi-graphs end -->

<!-- readthedocs-preview anemoi-models start -->
----
📚 Documentation preview 📚: https://anemoi-models--464.org.readthedocs.build/en/464/

<!-- readthedocs-preview anemoi-models end -->

JesperDramsch commented:
Yes exactly, so it's much easier to say "hey I have this checkpoint in an s3 bucket" or similar.

mchantry commented:
@JesperDramsch thanks for the contribution. Is there an active use-case for the GCS loading? There is an active user-base for azure, so I think implementing the azure one would be a nice addition, but wonder if you could save yourself some work by leaving GCS for when a use-case arises?

JesperDramsch committed:
fix(checkpoint): phase 1 bugfixes from loading-orchestration

Includes detect_format_from_data() needed by source base class,
update_metadata() on CheckpointContext, and other fixes.
These 6 files have 283 insertions/137 deletions vs main.

JesperDramsch committed:
feat(checkpoint): sources g1-g4 — Local, HTTP, S3 sources (#458)

Cherry-picked from loading-orchestration with Phase 1 fixes.

Sources:
- CheckpointSource(PipelineStage) base class
- LocalSource: torch.load with format detection
- HTTPSource: async download via download_with_retry
- S3Source: lazy boto3, URL parsing, temp file cleanup

Includes gate tests for all source implementations (17/17 passing).

JesperDramsch committed:
fix(checkpoint): align Phase 1 tests with implementation (#458)

Fix test-implementation mismatches in test_formats.py and test_utils.py:
- Unknown/corrupted/empty files raise errors, not return defaults
- validate_checkpoint NaN/Inf tests pass check_tensors=True
- Network test exception handlers catch CheckpointSourceError

Sources-g5 gate passing: 151+ tests green, 5 skipped (network).

JesperDramsch committed:
fix(checkpoint): add boto3 as optional s3 dependency for CI tests (#458)

S3 source tests fail on CI because boto3/botocore are not installed.
Add optional-dependencies.s3 with boto3 and include it in test extras.

JesperDramsch committed:
refactor(checkpoint): remove gate scaffolding from source tests (#458)

Replace sprint gate docstrings with standard module docstrings.

anaprietonem commented on training/src/anemoi/training/checkpoint/pipeline.py:
```diff
@@ -39,7 +39,7 @@
 ...         {'_target_': 'anemoi.training.checkpoint.sources.LocalSource',
 ...          'path': '/tmp/checkpoint.pt'},
-...         {'_target_': 'anemoi.training.checkpoint.loaders.WeightsOnlyLoader',
+...         {'_target_': 'anemoi.training.checkpoint.loading.LoadingStrategy',
```
I assume this is coming from the changes in the loaders, but this won't work on this PR?

JesperDramsch replied:
Done in 1c4761611. The `from_config` docstring example no longer references `LoadingStrategy` from the `loading` package (which is not present on this PR's branch).

anaprietonem reviewed (changes requested):
Nice work Jesper, very nice to start having this functionality! left some comments for changes and some questions to clarify

JesperDramsch commented on training/tests/unit/checkpoint/sources/test_s3.py:
```diff
@@ -0,0 +1,83 @@
+@pytest.mark.asyncio
+async def test_s3_source_parses_s3_url() -> None:
+    """s3://bucket/path/to/model.ckpt -> bucket='bucket', key='path/to/model.ckpt'."""
+    source = S3Source(url="s3://my-bucket/checkpoints/model.ckpt")
```
That would be great actually.

JesperDramsch replied:
Added as `test_s3_source_downloads_from_real_public_bucket` in 1c4761611 — a `@pytest.mark.network`-gated test that downloads from the public NOAA GHCN bucket via the real `anemoi.utils.remote.s3.download_file` path. Skips cleanly when `obstore` or `anemoi-utils[s3]` are not installed. If there is an internal team bucket you would prefer over the public NOAA one, happy to swap it in.

JesperDramsch commented on training/pyproject.toml:
```diff
@@ -91,13 +91,16 @@ optional-dependencies.profile = [
   "aiohttp>=3.8",
 ]
+optional-dependencies.s3 = [
+  "boto3>=1.28",
```
I will check. Best to re-use dependencies in my opinion.

JesperDramsch replied:
I have now completely switched the source to use the implementation in `anemoi-utils` thank you.

JesperDramsch replied:
Fixed in 1c4761611 (and the underlying boto3→obstore swap in fa4095953). `[s3]` extra now declares `anemoi-utils[s3]>=0.4.37`, which pulls obstore. End-to-end verified against the public NOAA S3 bucket.

JesperDramsch commented on training/src/anemoi/training/checkpoint/formats.py:
```diff
@@ -518,17 +544,32 @@ def convert_lightning_to_pytorch(
 
 
-def is_format_available(checkpoint_format: Literal["lightning", "pytorch", "state_dict"]) -> bool:  # noqa: ARG001
+def is_format_available(checkpoint_format: Literal["lightning", "pytorch", "safetensors", "state_dict"]) -> bool:
```
Oh yeah valid, this slipped in when I copied my old prototype, sorry.

JesperDramsch replied:
Done in 1c4761611. All safetensors references removed from `formats.py` (and the `Literal[...]` type alias no longer includes it). Validation utilities also no longer probe for the safetensors module.

JesperDramsch commented on training/src/anemoi/training/checkpoint/sources/base.py:
```diff
@@ -0,0 +1,129 @@
+"""Abstract base class for checkpoint sources.
+
+Checkpoint sources are responsible for acquiring checkpoint data from
+various locations (local filesystem, S3, HTTP, etc.) and populating
```
It will have to be user-supplied.

JesperDramsch replied:
Documented in 1c4761611. Added a paragraph at the top of `sources/base.py` clarifying that source selection is explicit via Hydra `_target_`. 

There is no scheme-based dispatcher that automatically routes `s3://` to `S3Source`. Each source has a `supports(...)` classmethod but it is opt-in validation, not automatic dispatch. The motivation is that someone could legitimately want `LocalSource` against a FUSE-mounted S3 path without the pipeline second-guessing the scheme.

anaprietonem replied:
Thanks Jesper, for that particular example what would be the error raise if someone tries to use a s3:// path while instantiating a LocalSource?

JesperDramsch replied:
If someone configures `LocalSource` with `s3://bucket/key`, the path gets interpreted as a literal local path and `Path("s3://bucket/key").resolve()` won't exist on disk, so you get a `CheckpointNotFoundError`. Added that to the module docstring with a note that callers wanting scheme-based routing should dispatch on `urlparse(uri).scheme` themselves before picking a `_target_`.

JesperDramsch committed:
refactor(checkpoint): use anemoi-utils s3_client instead of raw boto3 (#458)

Replace direct boto3.client("s3") with anemoi.utils.remote.s3.s3_client(),
which provides per-bucket endpoint/credential config from settings.toml,
thread-safe client caching, and ECMWF object-store compatibility.

JesperDramsch committed:
checkpoint: implement validation module for pipeline health

Adds training/src/anemoi/training/checkpoint/validation.py with:
- CheckpointPipelineValidator.validate_environment_setup()
- CheckpointPipelineValidator.validate_configuration()
- validate_pipeline_health()

Replaces the ImportError sentinel fallback in checkpoint/__init__.py
with a direct import now that the module exists. Restores the
pre-execution validation hook in pipeline.py, which silently degraded
to validation_skipped="module_unavailable" before.

JesperDramsch committed:
checkpoint(sources): switch S3Source to anemoi-utils obstore backend

Replace boto3-based download with anemoi.utils.remote.s3.download_file,
which is obstore-backed. boto3 is no longer a runtime or test dependency.

- training/src/anemoi/training/checkpoint/sources/s3.py: call
  download_file(url, target, overwrite, resume, verbosity) directly;
  drop the s3_client + region constructor parameter. Map
  FileNotFoundError -> CheckpointNotFoundError, ImportError (obstore
  missing) and (OSError, ValueError, RuntimeError) -> CheckpointSourceError.
  Install hint references the correct extra: anemoi-utils[s3].
- training/tests/checkpoint/sources/test_s3.py: inject a stub
  anemoi.utils.remote.s3 module via patch.dict(sys.modules, ...). New
  test for ImportError-at-call-time (obstore missing). Drop botocore
  ClientError mock.
- training/pyproject.toml: [s3] extra now anemoi-utils[s3]>=0.4.37
  instead of boto3>=1.28. anemoi-utils[s3] pulls obstore.

JesperDramsch committed:
checkpoint(sources): address PR #464 review feedback

Resolves nine of the ten inline comments from @anaprietonem on PR #464,
plus the @mchantry GCS suggestion.

Code changes:
- formats.py: remove all safetensors references (anemoi does not use it
  yet) and harden Lightning checkpoint detection. The version key
  'pytorch-lightning_version' is now the primary marker; a 2+
  auxiliary-key match is the fallback so a future Lightning rename of
  one key doesn't break detection silently.
- pipeline.py: docstring example no longer references LoadingStrategy
  from the loading package (not present on this PR's branch).
- sources/base.py: documents that source selection is explicit via
  Hydra _target_, not scheme-inferred.
- validation.py / test_validation.py: cherry-picked from PR #998 so
  validate_pipeline_health actually does something instead of being a
  raising sentinel. Drops the safetensors info line for consistency.
- tests/checkpoint/conftest.py: pulls sample_checkpoint up from the
  sources sub-conftest so other layers can use it. sources/conftest.py
  is removed (was a single fixture).
- tests/.../test_s3.py: adds a @pytest.mark.network integration test
  that downloads from the public NOAA GHCN-Daily bucket via S3Source.
  Skips cleanly when obstore or anemoi-utils[s3] are unavailable.

Phase 1 caveat:
This commit modifies pipeline.py and formats.py, both listed in
sprint/BOUNDARIES.md as Phase 1 frozen. The changes are direct
responses to reviewer requests on PR #464 (Ana's comments at
pipeline.py:42 and formats.py:51), so they ride along with this PR
rather than blocking it on a separate Phase 1 patch.

GCS scope: @mchantry suggested dropping GCS and prioritising Azure.
GCS stays out of scope; an AzureSource follow-up issue will be filed
separately.

JesperDramsch commented on training/tests/checkpoint/sources/conftest.py:
```diff
@@ -0,0 +1,15 @@
+
+
+@pytest.fixture
+def sample_checkpoint(tmp_path: Path) -> Path:
```
Done in 1c4761611. `sample_checkpoint` now lives in `training/tests/checkpoint/conftest.py` so other layers can reuse it. The single-fixture `sources/conftest.py` was removed.

JesperDramsch commented on training/src/anemoi/training/checkpoint/formats.py:
```diff
@@ -26,12 +26,68 @@
+        return "pytorch"
+
+    # Check for Lightning-specific keys (exclude generic training state keys)
+    lightning_specific_keys = {
```
Hardened in 1c4761611. The `pytorch-lightning_version` key is now the primary marker, and the auxiliary key set falls back to a 2+ match so detection survives a single Lightning rename. Added a docstring note about the tested PTL/torch range. Existing format tests still pass.

JesperDramsch commented on training/src/anemoi/training/checkpoint/catalog.py:
```diff
@@ -285,7 +285,7 @@ def _get_sources(cls) -> dict[str, str]:
         """Get the registry of loaders, discovering if needed."""
         if cls._loaders is None:
-            cls._loaders = cls._discover_components("anemoi.training.checkpoint.loaders", "LoadingStrategy")
+            cls._loaders = cls._discover_components("anemoi.training.checkpoint.loading", "LoadingStrategy")
```
No code change, as the design is intentional. `_get_loaders` calls `_discover_components("anemoi.training.checkpoint.loading", "LoadingStrategy")`, which returns an empty dict + warning if the `loading` package is missing (see `catalog.py:280-298`). 

So on this branch the catalog reports zero loaders and zero modifiers without failing and once PR #998 (loading) and PR #442 (modifiers) merge into main, the same call discovers them automatically. 

The pipeline/`from_config` docstring example at `pipeline.py:42` was the only place this assumption leaked into user-visible code; that has been fixed in 1c4761611.

anaprietonem replied:
thanks for the clarification Jesper. Could you please add a comment in the code explaining that's the expected behaviour and that this functionality will be extended with PRs 998 and 442? So if users look at this they understand that's work in progress.

JesperDramsch replied:
Done, added a docstring to both `_get_loaders` and `_get_modifiers` explaining that the packages are populated by PRs #998 and #442, and that an empty registry with a discovery warning is the expected behaviour until those land.

JesperDramsch commented on training/src/anemoi/training/checkpoint/__init__.py:
```diff
@@ -46,8 +46,17 @@
 try:
     from .validation import validate_pipeline_health
```
Working now. The previous code was a `try/except ImportError` sentinel because `checkpoint/validation.py` had never been written. Cherry-picked the implementation from PR #998 in 1e0f008ba. `validate_pipeline_health(context)` now does three things:
- Checks every `stage_{i}_{name}` metadata key is a completion marker (no `"failed:"` prefix).
- If any source stage ran, asserts `context.model.weights_initialized is True` so a swallowed loader error does not silently produce a random-weight model.
- Verifies structural invariants on the context (no optimizer without model, no `pl_module` with a non-Lightning format, etc.) and surfaces any error reported by the pre-execution environment/config validation.

API mirrors `validate_checkpoint` in `utils.py` — returns `True` or raises `CheckpointValidationError(.validation_errors=[...])`. Tests at `training/tests/checkpoint/test_validation.py` (21 tests, all green).

JesperDramsch commented on training/src/anemoi/training/checkpoint/sources/s3.py:
```diff
@@ -0,0 +1,271 @@
+LOGGER = logging.getLogger(__name__)
+
+
+class S3Source(CheckpointSource):
```
Done in fa4095953. `S3Source` now calls `anemoi.utils.remote.s3.download_file(url, target, overwrite, resume, verbosity)` directly (obstore-backed). No boto3 client construction in this PR. The `pyproject.toml` `[s3]` extra was switched to `anemoi-utils[s3]>=0.4.37`, which pulls obstore transitively. boto3 and botocore are no longer dependencies.

JesperDramsch committed:
Merge remote-tracking branch 'origin/main' into feature/checkpoint-sources

# Conflicts:
#	training/pyproject.toml
#	training/tests/unit/checkpoint/sources/__init__.py
#	training/tests/unit/checkpoint/sources/test_base.py
#	training/tests/unit/checkpoint/sources/test_http.py
#	training/tests/unit/checkpoint/sources/test_local.py
#	training/tests/unit/checkpoint/sources/test_s3.py
#	training/tests/unit/checkpoint/test_utils.py
#	training/tests/unit/checkpoint/test_validation.py

JesperDramsch committed:
checkpoint(sources): split anemoi settings into secrets file for CI

anemoi-utils rejects access_key_id / secret_access_key keys in
settings.toml and requires them in settings.secrets.toml even when
they are empty reset values. CI test failure:
  SystemError: Configuration file ... settings.toml should not hold
  entries {'secret_access_key', 'access_key_id'}.

Fix splits the per-bucket override into the two files so the
@pytest.mark.network integration test (real NOAA S3 download) passes.

JesperDramsch committed:
checkpoint(sources): chmod 600 the anemoi secrets file in CI test

anemoi-utils refuses to read settings.secrets.toml unless it is mode
0600 (world-not-readable). CI failure:
  SystemError: Configuration file ... settings.secrets.toml is not
  secure. Please run `chmod 600 ...`.

Fix sets the mode immediately after writing.

anaprietonem commented on training/src/anemoi/training/checkpoint/__init__.py:
```diff
@@ -46,8 +46,17 @@
+        msg = (
+            "validate_pipeline_health is not available because the "
+            "validation module could not be imported. Ensure that "
+            "anemoi.training.checkpoint.validation is installed."
```
why this error message? (as in this should be installed whenever someone does `pip install anemoi-training` right) (there is no separate checkpoint package(?))

JesperDramsch replied:
Good catch, that fallback was misleading. It's gone now. The validation module ships with anemoi-training so there's no reason for it to be missing at import time. Silly leftover from the gating logic before.

JesperDramsch committed:
docs(checkpoint): clarify catalog discovery scope and LocalSource s3 behaviour

Addresses PR #464 review feedback:

- catalog.py: explain in _get_loaders and _get_modifiers docstrings that
  the loading and modifiers packages are populated by anemoi-core PRs
  #998 and #442 respectively; until those land, discovery returns an
  empty registry and the warning is expected behaviour.
- sources/base.py: extend the source-selection paragraph with a concrete
  example of what happens when LocalSource receives an s3:// URI
  (CheckpointNotFoundError, because the URI is interpreted as a literal
  local path).

anaprietonem committed:
Merge branch 'main' into feature/458-refactor-restart-of-training-warm-start-forked-run

anaprietonem reviewed (approved):
Thanks Jesper for addressing the comments and implementing this feature! nice work!
